doc/guides/smf/index.rst

```diff
@@ -0,0 +1,267 @@
+.. _smf:
+
+State Machine Framework
+#######################
+
+Overview
+========
+
+The State Machine Framework (SMF) is an application agnostic framework that
+provides an easy way for developers to integrate state machines into their
+application. The framework can be added to any project by including the
+CONFIG_SMF=y into the projects prj.conf file.
+
+State Creation
+==============
+
+A state is represented by three functions, where one function implements the
+Entry actions, another function implements the Run actions, and the last
+function implements the Exit actions. The prototype for these functions is as
+follows: **void funct(void *obj)**, where the **obj** parameter is a user
+defined structure that has the state machine context, **struct smf_ctx**, as
+its first member. For example::
+
+	struct user_object {
+ 		struct smf_ctx ctx;
+ 		/* All User Defined Data Follows */
+	};
+
+The **struct smf_ctx** member must be first because the state machine
+framework's functions casts the user defined object to the **struct smf_ctx**
+type with the following macro: **SMF_CTX(o)**
+
+For example instead of doing this **(struct smf_ctx *)&user_obj**, you could
+use **SMF_CTX(user_obj)**.
+
+The following macros can be used for easy state creation:
+    - HSM_CREATE_MSG_STATE
+    - HSM_CREATE_STATE
+    - FSM_CREATE_MSG_STATE
+    - FSM_CREATE_BASIC_STATE
+
+**HSM_CREATE_MSG_STATE(me, mx, e, r, x, p)** is used to create a hierarchical
+state that can have an optional parent state and prints a message in the
+optional Entry action and a message in the optional Exit action. The macro has
+the following parameters:
```

Comment:
Note that the parameters are now out of date.  Probably better to document the parameters in the code only and just provide a summary of the macro and it's intended use here.

Reply:
I would suggest looking into the `:c:macro:` role and using it as needed to link to the API docs for this instead.

Reply:
Thanks for the suggestion